sdk/src/signers.rs

```diff
@@ -2,10 +2,13 @@ use crate::{
     pubkey::Pubkey,
     signature::{Signature, Signer},
 };
+use std::error;
 
 pub trait Signers {
     fn pubkeys(&self) -> Vec<Pubkey>;
+    fn try_pubkeys(&self) -> Result<Vec<Pubkey>, Box<dyn error::Error>>;
     fn sign_message(&self, message: &[u8]) -> Vec<Signature>;
+    fn try_sign_message(&self, message: &[u8]) -> Result<Vec<Signature>, Box<dyn error::Error>>;
```

Comment:
I must not have describe the issue here clearly enough.
These are all the errors RemoteKeypair could return on signing: https://github.com/solana-labs/solana/blob/ab361a807312f742174ebe78546bab995b568738/remote-wallet/src/remote_wallet.rs#L18
I can't convert these to some kind of native sdk error (due to crate circular-dependencies) unless I move RemoteWalletError into solana-sdk or somewhere else on which sdk and remote-wallet can both depend.

I suppose I could create one mega SigningError enum that includes all the possible errors from all the current Signers, but that seems hard to maintain (and antithetical to the work you've done getting rid of the common result). It would also bring an hidapi crate dependency into sdk.

Reply:
The point of a trait is to provide a consistent interface between multiple implementations. If every implementation of the trait can potentially return a different set of errors, how are its users supposed to work with that?

I think SigningError can be a lot bigger.  It doesn't necessary need hidapi, but could duplicate any of those errors if you think the user would be able to do anything with that data.  And for the remaining, sure, a catch-all error with a string.

`Box<dyn Error>` should only be used in quick-and-dirty code or by a top-level entrypoint, when the only thing you'd do with that error is print it.

Reply:
Okay, I think I've finally connected the mental dots on this. See what you think.